Crash loop when enabling biometrics

### What happened?

Not quite actually, but when I tap "Enable" when biometrics have already been set beforehand, the app crashes in loop

### App Version

6.4.0

### Android Version

29

### (Optional) Relevant log output

```shell
java.lang.RuntimeException: Exception while computing database live data.at androidx.room.v.run(SourceFile:274)at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)at java.lang.Thread.run(Thread.java:919)Caused by: net.sqlcipher.database.SQLiteException: file is not a database: , while compiling: select count(*) from sqlite_master;at net.sqlcipher.database.SQLiteCompiledSql.native_compile(Native Method)at net.sqlcipher.database.SQLiteCompiledSql.compile(SourceFile:16)at net.sqlcipher.database.SQLiteCompiledSql.<init>(SourceFile:31)at net.sqlcipher.database.SQLiteProgram.<init>(SourceFile:106)at net.sqlcipher.database.SQLiteQuery.<init>(SourceFile:1)at net.sqlcipher.database.SQLiteDirectCursorDriver.query(SourceFile:9)at net.sqlcipher.database.SQLiteDatabase.rawQueryWithFactory(SourceFile:37)at net.sqlcipher.database.SQLiteDatabase.rawQuery(SourceFile:1)at net.sqlcipher.database.SQLiteDatabase.keyDatabase(SourceFile:31)at net.sqlcipher.database.SQLiteDatabase.openDatabaseInternal(SourceFile:4)at net.sqlcipher.database.SQLiteDatabase.openDatabase(SourceFile:10)at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SourceFile:8)at net.sqlcipher.database.SQLiteOpenHelper.getWritableDatabase(SourceFile:14)at net.sqlcipher.database.SupportHelper.getWritableDatabase(SourceFile:6)at androidx.room.q.b(SourceFile:5)at androidx.room.q.n(SourceFile:9)at a.b.x(SourceFile:12)at com.philkes.notallyx.data.dao.g.call(SourceFile:243)at androidx.room.v.run(SourceFile:261)... 3 more
```

--- a/app/src/main/java/com/philkes/notallyx/data/NotallyDatabase.kt
+++ b/app/src/main/java/com/philkes/notallyx/data/NotallyDatabase.kt
@@ -2,6 +2,7 @@ package com.philkes.notallyx.data
 
 import android.app.Application
 import android.os.Build
+import androidx.annotation.RequiresApi
 import androidx.lifecycle.Observer
 import androidx.room.Database
 import androidx.room.Room
@@ -21,6 +22,7 @@ import com.philkes.notallyx.presentation.viewmodel.preference.BiometricLock
 import com.philkes.notallyx.presentation.viewmodel.preference.NotallyXPreferences
 import com.philkes.notallyx.presentation.viewmodel.preference.observeForeverSkipFirst
 import com.philkes.notallyx.utils.IO.getExternalMediaDirectory
+import com.philkes.notallyx.utils.security.SQLCipherUtils
 import com.philkes.notallyx.utils.security.getInitializedCipherForDecryption
 import java.io.File
 import net.sqlcipher.database.SupportFactory
@@ -103,12 +105,22 @@ abstract class NotallyDatabase : RoomDatabase() {
                     .addMigrations(Migration2, Migration3, Migration4, Migration5, Migration6)
             if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                 if (preferences.biometricLock.value == BiometricLock.ENABLED) {
-                    val initializationVector = preferences.iv.value!!
-                    val cipher = getInitializedCipherForDecryption(iv = initializationVector)
-                    val encryptedPassphrase = preferences.databaseEncryptionKey.value
-                    val passphrase = cipher.doFinal(encryptedPassphrase)
-                    val factory = SupportFactory(passphrase)
-                    instanceBuilder.openHelperFactory(factory)
+                    if (
+                        SQLCipherUtils.getDatabaseState(app, getCurrentDatabaseName(app)) ==
+                            SQLCipherUtils.State.ENCRYPTED
+                    ) {
+                        initializeDecryption(preferences, instanceBuilder)
+                    } else {
+                        preferences.biometricLock.save(BiometricLock.DISABLED)
+                    }
+                } else {
+                    if (
+                        SQLCipherUtils.getDatabaseState(app, getCurrentDatabaseName(app)) ==
+                            SQLCipherUtils.State.ENCRYPTED
+                    ) {
+                        preferences.biometricLock.save(BiometricLock.ENABLED)
+                        initializeDecryption(preferences, instanceBuilder)
+                    }
                 }
                 val instance = instanceBuilder.build()
                 if (observePreferences) {
@@ -143,6 +155,19 @@ abstract class NotallyDatabase : RoomDatabase() {
                 return instance
             }
             return instanceBuilder.build()
+        }
+
+        @RequiresApi(Build.VERSION_CODES.M)
+        private fun initializeDecryption(
+            preferences: NotallyXPreferences,
+            instanceBuilder: Builder<NotallyDatabase>,
+        ) {
+            val initializationVector = preferences.iv.value!!
+            val cipher = getInitializedCipherForDecryption(iv = initializationVector)
+            val encryptedPassphrase = preferences.databaseEncryptionKey.value
+            val passphrase = cipher.doFinal(encryptedPassphrase)
+            val factory = SupportFactory(passphrase)
+            instanceBuilder.openHelperFactory(factory)
         }
 
         object Migration2 : Migration(1, 2) {

--- a/app/src/main/java/com/philkes/notallyx/presentation/activity/main/fragment/settings/PreferenceBindingExtensions.kt
+++ b/app/src/main/java/com/philkes/notallyx/presentation/activity/main/fragment/settings/PreferenceBindingExtensions.kt
@@ -82,6 +82,9 @@ fun PreferenceBinding.setup(
             .setSingleChoiceItems(entries, checked) { dialog, which ->
                 dialog.cancel()
                 val newValue = enumEntries[which]
+                if (newValue == value) {
+                    return@setSingleChoiceItems
+                }
                 if (newValue == BiometricLock.ENABLED) {
                     when (context.canAuthenticateWithBiometrics()) {
                         BiometricManager.BIOMETRIC_SUCCESS -> onEnableSuccess()

--- a/app/src/main/java/com/philkes/notallyx/presentation/activity/note/EditActivity.kt
+++ b/app/src/main/java/com/philkes/notallyx/presentation/activity/note/EditActivity.kt
@@ -65,9 +65,9 @@ import com.philkes.notallyx.utils.changehistory.ChangeHistory
 import com.philkes.notallyx.utils.mergeSkipFirst
 import com.philkes.notallyx.utils.observeSkipFirst
 import com.philkes.notallyx.utils.wrapWithChooser
+import java.io.File
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
-import java.io.File
 
 abstract class EditActivity(private val type: Type) :
     LockedActivity<ActivityEditBinding>(), AddActions, MoreActions {
@@ -85,8 +85,8 @@ abstract class EditActivity(private val type: Type) :
     internal val notallyModel: NotallyModel by viewModels()
     internal lateinit var changeHistory: ChangeHistory
 
-    protected lateinit var undo: View
-    protected lateinit var redo: View
+    protected var undo: View? = null
+    protected var redo: View? = null
 
     override fun finish() {
         lifecycleScope.launch(Dispatchers.Main) {
@@ -256,12 +256,8 @@ abstract class EditActivity(private val type: Type) :
     protected open fun initChangeHistory() {
         changeHistory =
             ChangeHistory().apply {
-                canUndo.observe(this@EditActivity) { canUndo ->
-                    undo.isEnabled = canUndo 
-                }
-                canRedo.observe(this@EditActivity) { canRedo ->
-                    redo.isEnabled = canRedo 
-                }
+                canUndo.observe(this@EditActivity) { canUndo -> undo?.isEnabled = canUndo }
+                canRedo.observe(this@EditActivity) { canRedo -> redo?.isEnabled = canRedo }
             }
     }
 
@@ -304,9 +300,7 @@ abstract class EditActivity(private val type: Type) :
             search.prevMenuItem?.isEnabled = hasResults
         }
 
-        search.resultPos.observeSkipFirst(this) {
-            pos -> selectSearchResult(pos)
-        }
+        search.resultPos.observeSkipFirst(this) { pos -> selectSearchResult(pos) }
 
         binding.EnterSearchKeyword.apply {
             doAfterTextChanged { text ->
@@ -322,11 +316,12 @@ abstract class EditActivity(private val type: Type) :
         val amount = highlightSearchResults(query)
         this.search.results.value = amount
         if (amount > 0) {
-            search.resultPos.value = when {
-                amountBefore < 1 -> 0
-                search.resultPos.value >= amount -> amount - 1
-                else -> search.resultPos.value
-            }
+            search.resultPos.value =
+                when {
+                    amountBefore < 1 -> 0
+                    search.resultPos.value >= amount -> amount - 1
+                    else -> search.resultPos.value
+                }
         }
     }
 
@@ -407,24 +402,25 @@ abstract class EditActivity(private val type: Type) :
         }
         binding.BottomAppBarCenter.apply {
             removeAllViews()
-            undo = addIconButton(R.string.undo, R.drawable.undo, marginStart = 2) {
-                try {
-                        changeHistory.undo()
-                } catch (e: ChangeHistory.ChangeHistoryException) {
-                    Operations.log(application, e)
-                }
-            }
+            undo =
+                addIconButton(R.string.undo, R.drawable.undo, marginStart = 2) {
+                        try {
+                            changeHistory.undo()
+                        } catch (e: ChangeHistory.ChangeHistoryException) {
+                            Operations.log(application, e)
+                        }
+                    }
                     .apply { isEnabled = changeHistory.canUndo.value }
 
-            redo = addIconButton(R.string.redo, R.drawable.redo, marginStart = 2) {
-                try {
-                        changeHistory.redo()
-                } catch (e: ChangeHistory.ChangeHistoryException) {
-                    Operations.log(application, e)
-                }
-            }
+            redo =
+                addIconButton(R.string.redo, R.drawable.redo, marginStart = 2) {
+                        try {
+                            changeHistory.redo()
+                        } catch (e: ChangeHistory.ChangeHistoryException) {
+                            Operations.log(application, e)
+                        }
+                    }
                     .apply { isEnabled = changeHistory.canRedo.value }
-
         }
         binding.BottomAppBarRight.apply {
             removeAllViews()
@@ -831,7 +827,7 @@ abstract class EditActivity(private val type: Type) :
         var prevMenuItem: MenuItem? = null,
         var nextMenuItem: MenuItem? = null,
         var resultPos: NotNullLiveData<Int> = NotNullLiveData(-1),
-        var results: NotNullLiveData<Int> = NotNullLiveData(-1)
+        var results: NotNullLiveData<Int> = NotNullLiveData(-1),
     )
 
     companion object {
@@ -841,5 +837,4 @@ abstract class EditActivity(private val type: Type) :
         const val FOLDER_FROM = "folder_from"
         const val FOLDER_TO = "folder_to"
     }
-
 }

